Eject react and fix build bug

timizuoebideri1 opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes: #4083 
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->


# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

timizuoebideri1 commented on nautobot/ui/jest.config.js:
```diff
@@ -0,0 +1,43 @@
+const nautobotAlias = require("./aliases")
```
To implement nautobotAlias, I extract the jest configuration from package.json and paste it here.

timizuoebideri1 commented on nautobot/ui/scripts/build.js:
File created as a result of eject

timizuoebideri1 commented on nautobot/ui/scripts/start.js:
File created as a result of eject

timizuoebideri1 commented on nautobot/ui/scripts/test.js:
File created as a result of eject

timizuoebideri1 commented on nautobot/ui/config/webpack.config.js:
- implement `nautobotAlias`. Line `31` & `327`
- Port config from craco over to `web pack `

glennmatthews commented:
First thought is "oof, that's a lot of new boilerplate to ignore" :-( But may be necessary... taking a look now.

glennmatthews reviewed (commented):
So which files or functions are *not* unchanged boilerplate from craco/create-react-app? Where should I be focusing my review?

bryanculver commented on nautobot/ui/package.json:
```diff
@@ -62,13 +106,19 @@
         "eslint-config-react-app": "^7.0.1",
         "markdownlint-cli": "^0.35.0",
         "prettier": "^2.8.8",
         "react-app-alias": "^2.2.2",
         "react-app-alias-ex": "^2.1.0",
```
It seems like you're no longer using these libraries (it appears you have your own alias tool) so these could probably be deleted.

bryanculver commented on nautobot/ui/package.json:
```diff
@@ -14,33 +15,76 @@
+        "style-loader": "^3.3.1",
         "swr": "^1.3.0",
-        "web-vitals": "^2.1.4"
+        "tailwindcss": "^3.0.2",
+        "terser-webpack-plugin": "^5.2.5",
+        "web-vitals": "^2.1.4",
+        "webpack": "^5.64.4",
+        "webpack-dev-server": "^4.6.0",
```
We don't use tailwind and it seems that CRA has dev dependencies as direct dependencies (probably because if it didn't they would never get installed, as they would show up a dev dependences of react-scripts/cra).

Can we quickly review things and move the obvious ones down to dev dependencies?

bryanculver reviewed (commented):
Seems somewhat reasonable and we'll improve it over time but I think there are two quick cleanups we could make with ejecting and I'd be comfortable approving this.

timizuoebideri1 commented:
> So which files or functions are _not_ unchanged boilerplate from craco/create-react-app? Where should I be focusing my review?

> So which files or functions are _not_ unchanged boilerplate from craco/create-react-app? Where should I be focusing my review?

1. `nautobot/ui/aliases.js`

 2. moved from `craco.config.js` to `webpack.config.js`:
    moved `webpackConfig.output.filename` to Line  222
    moved `config.output.publicPath` to Line 228
    moved `webpackConfig.output.assetModuleFilename` to 224\
    moved `webpackConfig.output.chunkFilename` to 223
    moved `webpackConfig.devtool` to 202
    moved `webpackConfig.plugins[5].options.filename` to 640
    moved `webpackConfig.devServer` to 206 - 208

glennmatthews commented on nautobot/ui/aliases.js:
```diff
@@ -0,0 +1,48 @@
+
+const createJestAliasesFromJsconfig = (jsconfig) => {
+    const webpackAliases = {};
+    // ^@generated/(.*)$
```
?

glennmatthews commented on nautobot/ui/aliases.js:
```diff
@@ -0,0 +1,48 @@
+};
+
+const createJestAliasesFromJsconfig = (jsconfig) => {
+    const webpackAliases = {};
```
Should `webpackAliases` be changed to `jestAliases` throughout this function?

glennmatthews commented:
> > So which files or functions are _not_ unchanged boilerplate from craco/create-react-app? Where should I be focusing my review?
> 
> > So which files or functions are _not_ unchanged boilerplate from craco/create-react-app? Where should I be focusing my review?
> 
>     1. `nautobot/ui/aliases.js`
> 
>     2. moved from `craco.config.js` to `webpack.config.js`:
>        moved `webpackConfig.output.filename` to Line  222
>        moved `config.output.publicPath` to Line 228
>        moved `webpackConfig.output.assetModuleFilename` to 224
>        moved `webpackConfig.output.chunkFilename` to 223
>        moved `webpackConfig.devtool` to 202
>        moved `webpackConfig.plugins[5].options.filename` to 640
>        moved `webpackConfig.devServer` to 206 - 208

It would be great to revisit each of these customizations and add more in-depth comments explaining what they're doing and why, maybe also add some really obvious comment like `// NAUTOBOT CHANGE TO CRACO DEFAULTS` to each one as well so that this can be easily located by a casual reader.

glennmatthews commented on nautobot/ui/config/webpack.config.js:
```diff
@@ -0,0 +1,779 @@
+        devtool: process.env.NAUTOBOT_DEBUG
+            ? "eval-source-map"
+            : "hidden-source-map",
+        // NAUTOBOT CHANGE TO CRACO DEFAULTS
+        // TODO(jathan): This might be safely deleted.
+        devServer: {
+            writeToDisk: true, // Write files to disk in dev mode so Django can serve the assets
+        },
```
Try removing this and see what breaks

glennmatthews replied:
removed in #4248.

glennmatthews commented on nautobot/ui/config/webpack.config.js:
```diff
@@ -0,0 +1,779 @@
+                        configOverwrite: {
+                            compilerOptions: {
+                                sourceMap: isEnvProduction
+                                    ? shouldUseSourceMap
+                                    : isEnvDevelopment,
+                                skipLibCheck: true,
+                                inlineSourceMap: false,
+                                declarationMap: false,
+                                noEmit: true,
+                                incremental: true,
+                                tsBuildInfoFile: paths.appTsBuildInfoFile,
+                            },
+                        },
+                        context: paths.appPath,
+                        diagnosticOptions: {
+                            syntactic: true,
+                        },
+                        mode: "write-references",
+                        // profile: true,
+                    },
+                    issue: {
+                        // This one is specifically to match during CI tests,
+                        // as micromatch doesn't match
+                        // '../cra-template-typescript/template/src/App.tsx'
+                        // otherwise.
+                        include: [
+                            { file: "../**/src/**/*.{ts,tsx}" },
+                            { file: "**/src/**/*.{ts,tsx}" },
+                        ],
+                        exclude: [
+                            { file: "**/src/**/__tests__/**" },
+                            { file: "**/src/**/?(*.){spec|test}.*" },
+                            { file: "**/src/setupProxy.*" },
+                            { file: "**/src/setupTests.*" },
+                        ],
+                    },
+                    logger: {
+                        infrastructure: "silent",
+                    },
+                }),
```
nuke?

glennmatthews commented on nautobot/ui/jest.config.js:
```diff
@@ -0,0 +1,47 @@
+
+module.exports = {
+    roots: ["<rootDir>/src"],
+    collectCoverageFrom: ["src/**/*.{js,jsx,ts,tsx}", "!src/**/*.d.ts"],
```
Remove ts/tsx refs throughout this file?

glennmatthews replied:
removed in #4248.

glennmatthews commented on nautobot/ui/package.json:
```diff
@@ -19,28 +19,34 @@
+        "identity-obj-proxy": "^3.0.0",
+        "prompts": "^2.4.2",
         "react": "^18.2.0",
+        "react-app-polyfill": "^3.0.0",
+        "react-dev-utils": "^12.0.1",
```
Check whether these various dependencies are needed for prod or whether dev-only is preferable?

glennmatthews commented on nautobot/ui/package.json:
```diff
@@ -19,28 +19,34 @@
         "react-router-dom": "^6.4.3",
-        "react-scripts": "5.0.1",
         "redux-persist": "^6.0.0",
+        "semver": "^7.3.5",
```
dev-dependency?

glennmatthews commented on nautobot/ui/scripts/build.js:
```diff
@@ -0,0 +1,224 @@
+                    process.env.CI.toLowerCase() !== "false") &&
+                messages.warnings.length
+            ) {
+                // Ignore sourcemap warnings in CI builds. See #8227 for more info.
```
Might want to look into directly running `npm build` in CI so that we get these warnings-as-errors checks?

glennmatthews commented on nautobot/ui/scripts/test.js:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+function isInMercurialRepository() {
+    try {
+        execSync("hg --cwd . root", { stdio: "ignore" });
+        return true;
+    } catch (e) {
+        return false;
+    }
+}
```
remove?

glennmatthews replied:
removed in #4248.

glennmatthews commented:
I'll work on addressing these comments.

glennmatthews commented on nautobot/ui/config/webpack.config.js:
```diff
@@ -0,0 +1,779 @@
+const sassRegex = /\.(scss|sass)$/;
+const sassModuleRegex = /\.module\.(scss|sass)$/;
+
+const hasJsxRuntime = (() => {
```
removed in #4248.
